proof/crefine/Move_C.thy

```diff
@@ -1119,45 +1133,54 @@ lemma updateObject_cte_tcb:
         return (KOTCB (updF (\<lambda>_. ctea) tcb))
     od)"
   using tc unfolding tcb_cte_cases_def
-  apply -
-  apply (clarsimp simp add: updateObject_cte Let_def
-    tcb_cte_cases_def objBits_simps' tcbSlots shiftl_t2n
-    split: if_split_asm cong: if_cong)
-  done
+  by (clarsimp simp: updateObject_cte Let_def tcb_cte_cases_neqs gen_objBits_simps tcbSlots
+               split: if_split_asm
+               cong: if_cong)
+
+lemma self_elim:
+  "\<And>P Q. \<lbrakk> P \<Longrightarrow> Q; P \<rbrakk> \<Longrightarrow> Q"
+  by blast
```

Comment:
Out of curiosity, do you remember why this is needed?

Reply:
It's needed in the two lemmas below, where we have that Isabellism of
```
[| blah ==> thesis; asm1; asm2 |] ==> thesis
```
and this rule is what I came up with to unscrew that. If there's a nicer way, please let me know. This pattern comes up when using `obtain` which is from Isar. The HOL way, we'd get an existential.